### PR TITLE
Alter group2invoke

### PR DESCRIPTION
**Pull request by calebmkim**
This PR alters the group2invoke pass to be able to handle a) guarded assignments and b) writes to combinational cells. 

Edit: Failing either because a) applying group2invoke now at places where it wasn't before and b) this is increasing the cycle count, I think because the increased combinational groups w/ the invoke statement are getting turned into regular groups thereby increasing the cycle count. 
Edit2: I overrode a bunch of test cases to get these to pass.

### CODE REVIEW

**Comment by rachitnigam** [comment]
Yeah, that's probably right. We should address this issue to get things to take the same number of cycles: https://github.com/cucapra/calyx/issues/911

**Comment by rachitnigam** [comment]
@calebmkim when you get a chance, can you resolve the conflicts on this branch and ping me for a review?

**Comment by calebmkim** [comment]
This will fail. Bc group2seq and group2invoke overlap/work together in their effects, I basically just went with the master's changes for all the conflicts, and am going to do `runt -d` to see what changes my group2invoke is making.

**Comment by calebmkim** [comment]
@rachitnigam couple of questions. 

Instead of doing the analysis on each `enable`, should I have a map that maps group names to invokes and then just clone from the map -- similar to `group2seq`, or should I keep it as is? 

Also, what should I do with something like 
```
group A {
    r.write_en = 1'd1; 
    r.in = r.out; 
    A[done] = r.done
}
```
Should this turn into `invoke r(in = r.out)()`? Right now I don't currently do that but can change the pass to do so.

**Comment by rachitnigam** [comment]
That might be a good idea!

**Comment by calebmkim** [comment]
@rachitnigam when you get the change could you take a look? Thanks.

File: calyx/src/passes/group_to_invoke.rs [hunk 1]
```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
     }
 }
 
+// Returns true if port's parent is cell. We can do this by checking if they
+// have the same name.
+fn cell_is_parent(port: &ir::Port, cell: &ir::RRC<ir::Cell>) -> bool {
+    port.get_parent_name() == cell.borrow().name()
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:40]: This should use [`Rc::ptr_eq`](https://doc.rust-lang.org/stable/std/rc/struct.Rc.html#method.ptr_eq) to check that the parent pointer of the port is the same as the provided cell. You can match on the `.parent` field and see if the `uprade()` call to the parent pointer generates the same pointer

File: calyx/src/passes/group_to_invoke.rs [hunk 2]
```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
...
         {
             let name = assign.dst.borrow().name.clone();
-            inputs.push((name, Rc::clone(&assign.src)));
+            if assign.guard.is_true() {
+                inputs.push((name, Rc::clone(&assign.src)));
+            } else {
+                // assign has a guard condition,so need a wire
+                let width = assign.dst.borrow().width;
+                let wire =
+                    builder.add_primitive("std_wire", "std_wire", &[width]);
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:86]: Let's use a more informative prefix than "std_wire" here. Maybe something like `guard`?

File: calyx/src/passes/group_to_invoke.rs [hunk 3]
```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
...
     for assign in assigns {
-        // If the cell's port is being used as a source, add the dst to
-        // outputs
-        if cell_is_parent(&assign.src.borrow())
-            && assign.src != comp.borrow().get_with_attr("done")
-        {
-            let name = assign.src.borrow().name.clone();
-            outputs.push((name, Rc::clone(&assign.dst)));
+        // We know that all assignments in this group should write to either a)
+        // a combinational component or b) comp or c) the group's done port
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:66]: Is this a precondition on the `assigns` array?

File: tests/passes/group2invoke/transform.expect [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  control {
+    seq {
+      invoke r(
+        in = 32'd1
+      )();
+      invoke r(
+        in = add.out
+      )()with comb_invoke;
```
Comment (rachitnigam) [inline tests/passes/group2invoke/transform.expect:34]: Argh, the printer does not add space before the `with` clause here. Can you fix this in `printer.rs`?

File: calyx/src/passes/group_to_invoke.rs [hunk 5]
```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
...
-        // Component must define a @go/@done interface
-        let cell = writes.pop().unwrap();
-        let maybe_go_port = cell.borrow().find_with_attr("go");
-        let maybe_done_port = cell.borrow().find_with_attr("done");
-        if maybe_go_port.is_none() || maybe_done_port.is_none() {
-            return Ok(Action::Continue);
-        }
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in groups.iter() {
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:130]: No need for the `iter` call here. Just do `for g in &groups`?

File: calyx/src/passes/group_to_invoke.rs [hunk 6]
```diff
@@ -31,100 +34,190 @@ impl Named for GroupToInvoke {
...
+            let group = g.borrow();
+            let mut writes: Vec<ir::RRC<ir::Cell>> =
+                ReadWriteSet::write_set(group.assignments.iter())
+                    .filter(|cell| match cell.borrow().prototype {
+                        ir::CellType::Primitive { is_comb, .. } => !is_comb,
+                        _ => true,
+                    })
+                    .collect_vec();
+            // should only have one write to a non-combinational component
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:139]: Why is this required? A `comb group` can have guarded assignments to a cell in it

**Review by rachitnigam — COMMENTED** [review]
Left a couple of comments and questions

File: calyx/src/passes/group_to_invoke.rs (hunk 3 above)
Comment (calebmkim) [reply calyx/src/passes/group_to_invoke.rs:66]: Yes, since before each call of this method, we check for this (among other things) and if it is not satisfied then we don't call the method and instead just `continue`

File: tests/passes/group2invoke.futil [hunk 7]
```diff
@@ -1,17 +0,0 @@
-// -p group2invoke -p dead-group-removal
```
[the remark this answers is not in the document]
Comment (calebmkim) [reply tests/passes/group2invoke.futil:1]: I broke it into 2 tests and put it in a folder.

File: calyx/src/passes/group_to_invoke.rs (hunk 6 above)
Comment (calebmkim) [reply calyx/src/passes/group_to_invoke.rs:139]: I meant "should only be 1 write, among writes to non-combinational components". I'll fix it to say "Excluding writes to combinational components, should only write to one cell".

**Comment by calebmkim** [comment]
@rachitnigam sorry to ping again but when you have time could you take a look?

File: calyx/src/passes/group_to_invoke.rs [hunk 8]
```diff
@@ -31,100 +34,200 @@ impl Named for GroupToInvoke {
     }
 }
 
+// Returns true if port's parent is cell. We can do this by checking if they
+// have the same name.
+fn cell_is_parent(port: &ir::Port, cell: &ir::RRC<ir::Cell>) -> bool {
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:39]: rename to `parent_is_cell`

File: calyx/src/passes/group_to_invoke.rs [hunk 9]
```diff
@@ -31,100 +34,200 @@ impl Named for GroupToInvoke {
...
+                inputs.push((name, Rc::clone(&assign.src)));
+            } else {
+                // assign has a guard condition,so need a wire
+                let width = assign.dst.borrow().width;
+                let wire = builder.add_primitive(
+                    name.clone().id + "_guarded_wire",
+                    "std_wire",
+                    &[width],
+                );
+                let wire_in_rrc = wire.borrow().get("in");
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:96]: No need for the `_rrc` suffix in these names

File: calyx/src/passes/group_to_invoke.rs [hunk 10]
```diff
@@ -31,100 +34,200 @@ impl Named for GroupToInvoke {
...
-        let maybe_go_port = cell.borrow().find_with_attr("go");
-        let maybe_done_port = cell.borrow().find_with_attr("done");
-        if maybe_go_port.is_none() || maybe_done_port.is_none() {
-            return Ok(Action::Continue);
-        }
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
+        let mut builder = ir::Builder::new(comp, sigs);
+        for g in &groups {
+            let group = g.borrow();
+            let mut writes: Vec<ir::RRC<ir::Cell>> =
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:141]: No need for an explicit type here

File: calyx/src/passes/group_to_invoke.rs [hunk 11]
```diff
@@ -31,100 +34,200 @@ impl Named for GroupToInvoke {
...
-        }
-
-        // Component must define a @go/@done interface
-        let cell = writes.pop().unwrap();
-        let maybe_go_port = cell.borrow().find_with_attr("go");
-        let maybe_done_port = cell.borrow().find_with_attr("done");
-        if maybe_go_port.is_none() || maybe_done_port.is_none() {
-            return Ok(Action::Continue);
-        }
+        let groups: Vec<ir::RRC<ir::Group>> = comp.groups.drain().collect();
```
Comment (rachitnigam) [inline calyx/src/passes/group_to_invoke.rs:137]:
```suggestion
        let groups = comp.groups.drain().collect_vec();
```

**Review by rachitnigam — APPROVED** [review]
Minor comments. Address and merge at your discretion!
